### ISSUE TITLE
Bug: [consistent-type-assertions] Omitting the default assertionStyle option causes a schema error

### ISSUE THREAD

**Issue by scottohara** [typescript-eslint/typescript-eslint#10463]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

The default options for the `@typescript-eslint/consistent-type-assertions` rule are:

```js
defaultOptions: [
  {
    assertionStyle: 'as',
    objectLiteralTypeAssertions: 'allow',
  },
],
```

If a user wishes to override just the `objectLiteralTypeAssertions` option to `'never'`, but retain the default `assertionStyle` option `'as'`, the expectation is that they could omit the `assertionStyle` option from their config, e.g.

```js
"@typescript-eslint/consistent-type-assertions": [
  "error",
  {
    objectLiteralTypeAssertions: 'never',
  },
],
```

Default rule options should be deeply merged with any user-supplied config, so the expected merged result would be:

```js
 {
    assertionStyle: 'as',
    objectLiteralTypeAssertions: 'never',
  },
```

However this results in the following error:

```
❯ eslint

Oops! Something went wrong! :(

ESLint: 9.16.0

Error: Key "rules": Key "@typescript-eslint/consistent-type-assertions":
        Value {"objectLiteralTypeAssertions":"never"} should NOT have additional properties.
                Unexpected property "objectLiteralTypeAssertions". Expected properties: "assertionStyle".
        Value {"objectLiteralTypeAssertions":"never"} should have required property 'assertionStyle'.
        Value {"objectLiteralTypeAssertions":"never"} should match exactly one schema in oneOf.
```

It would seem that the user-supplied config is completely replacing the default options instead of being merged?

### Reproduction Repository Link

https://github.com/scottohara/ts-eslint-consistent-type-assertions

### Repro Steps

1) clone the repo
2) `npm install`
3) `eslint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | `8.17.0` |
| `TypeScript`                           | `5.7.2` |
| `ESLint`                               | `9.16.0` |
| `node`                                 | `22.11.0` |

**Comment by bradzacher** [comment]
> Default rule options should be deeply merged with any user-supplied config

> The default option merging happens _after_ the schema check, not before. This happens because eslint does the schema check, but we apply default options. We cannot apply the option earlier because we cannot change eslint's behaviour! 

This is the same for all rules.

The schema for the rule is as indicated by the options type on the docs page:
https://typescript-eslint.io/rules/consistent-type-assertions/#options

If we wanted to allow you to apply partial options and let the default options apply for the missing required properties we would need to setup a more complex schema with the correct optional properties that align with the default. I'm not sure if that's worth doing in the general case...

It's worth noting that eslint recently released infra following our lead which allows rules to declare default options and have them work like you described. But we won't be able to rely on that until our minimum supported version aligns!

**Comment by bradzacher** [comment]
Cc @typescript-eslint/triage-team thoughts?

**Comment by scottohara** [comment]
Thanks @bradzacher , that does make sense.

> It's worth noting that eslint recently released infra following our lead which allows rules to declare default options and have them work like you described

Yes, that was what prompted me to start trimming my config to remove any redundant options that were already the default, across both core eslint rules and typescript-eslint rules; but I hadn’t appreciated that the schema checks were still handled first by eslint before merging in any TS-specific defaults.

(Perhaps by pure coincidence, this was the only rule in my config that failed the schema check. It must be the case that  any other rules that I’m providing partial config for have a more forgiving schema.)

**Comment by JoshuaKGoldberg** [comment]
> If we wanted to allow you to apply partial options and let the default options apply for the missing required properties we would need to setup a more complex schema with the correct optional properties that align with the default. I'm not sure if that's worth doing in the general case...

I'm lightly in favor of doing that in the general case. Unnecessarily complex ESLint configs are a _killer_ for public perception of ESLint and linters/linting in general. Anything we can do to let configs be simpler is a win IMO.

**Comment by JoshuaKGoldberg** [comment]
Talked a bit with @bradzacher - this isn't a super high priority thing, but we could take it in. @scottohara are you up for sending a PR?

**Comment by scottohara** [comment]
Sure thing. 

Leave this with me to do some digging in the code first to make sure I fully understand how things currently work.

Once I have that mental model, I’ll give some thought to how best to implement support for partial options and hopefully get a draft PR started.

**Comment by scottohara-frollo** [comment]
I spent some time looking into this and it would seem that we're actually pretty close already.

_If_ we accept a constraint that support for partial options requires eslint >= 9.15.0 (allowing us to lean on the infra work already done to support `meta.defaultOptions` in core), then the only reason why this doesn't work already is that rules in tseslint have historically stored their default options in `.defaultOptions` rather than one level down in `meta.defaultOptions`.

Further investigation shows that some work has already been done to support this, in https://github.com/typescript-eslint/typescript-eslint/pull/10339 , where the declared default options is hoisted out of `createRule`, and a reference to it duplicated in both locations.

I'm not sure the reasoning why that change was only applied to a particular subset of 14 rules, but I can confirm that applying the same change to the `consistent-type-assertions` rule does fix the validation error I was previously seeing when trying to use partial options as per above.

If we're OK with this approach, then one solution would be to simply extend the work in #10339 to all other rules.

Or, rather than have to touch every rule; is there a preference to instead modify the `createRule` function to do this automatically for all rules? e.g.

```ts
/* packages/utils/src/eslint-utils/RuleCreator.ts */

function createRule<
  Options extends readonly unknown[],
  MessageIds extends string,
  PluginDocs = unknown,
>({
  create,
  defaultOptions,
  meta,
}: Readonly<RuleWithMeta<Options, MessageIds, PluginDocs>>): RuleModule<
  MessageIds,
  Options,
  PluginDocs
> {
  return {
    create(context: Readonly<RuleContext<MessageIds, Options>>): RuleListener {
      const optionsWithDefault = applyDefault(defaultOptions, context.options);
      return create(context, optionsWithDefault);
    },
    defaultOptions,
    meta: {                      <---  was previously just `meta`
       ...meta,                  <---
       defaultOptions            <---
    },
  };
}
```

**Comment by scottohara** [comment]
I spent some time looking into this and it would seem that we're actually pretty close already.

_If_ we accept a constraint that support for partial options requires eslint >= 9.15.0 (allowing us to lean on the infra work already done to support `meta.defaultOptions` in core), then the only reason why this doesn't work already is that rules in tseslint have historically stored their default options in `.defaultOptions` rather than one level down in `meta.defaultOptions`.

Further investigation shows that some work has already been done to support this, in https://github.com/typescript-eslint/typescript-eslint/pull/10339 , where the declared default options is hoisted out of `createRule`, and a reference to it duplicated in both locations.

I'm not sure the reasoning why that change was only applied to a particular subset of 14 rules, but I can confirm that applying the same change to the `consistent-type-assertions` rule does fix the validation error I was previously seeing when trying to use partial options as per above.

If we're OK with this approach, then one solution would be to simply extend the work in #10339 to all other rules.

Or, rather than have to touch every rule; is there a preference to instead modify the `createRule` function to do this automatically for all rules? e.g.

```ts
/* packages/utils/src/eslint-utils/RuleCreator.ts */

function createRule<
  Options extends readonly unknown[],
  MessageIds extends string,
  PluginDocs = unknown,
>({
  create,
  defaultOptions,
  meta,
}: Readonly<RuleWithMeta<Options, MessageIds, PluginDocs>>): RuleModule<
  MessageIds,
  Options,
  PluginDocs
> {
  return {
    create(context: Readonly<RuleContext<MessageIds, Options>>): RuleListener {
      const optionsWithDefault = applyDefault(defaultOptions, context.options);
      return create(context, optionsWithDefault);
    },
    defaultOptions,
    meta: {                      <---  was previously just `meta`
       ...meta,                  <---
       defaultOptions            <---
    },
  };
}
```

**Comment by kirkwaiblinger** [comment]
drive-by:

> I'm not sure the reasoning why that change was only applied to a particular subset of 14 rules, but I can confirm that applying the same change to the `consistent-type-assertions` rule does fix the validation error I was previously seeing when trying to use partial options as per above.

The goal in #10339 was to solve the problem of extension rules crashing with eslint 9.15 (and urgently) - see https://github.com/typescript-eslint/typescript-eslint/issues/10338. It was not to generally adopt new ESLint's default options setup. Therefore, only extension rules were touched. As for why not _all_ extension rules even, see https://github.com/typescript-eslint/typescript-eslint/pull/10339#issuecomment-2480912245

**Comment by scottohara** [comment]
Thanks for the info, @kirkwaiblinger .

It sounds like I should look further into what makes rules like `prefer-destructuring` incompatible with the change.

**Comment by JoshuaKGoldberg** [comment]
Agreed, overhauling rules to use the new ESLint options defaults would be great but is definitely a bigger separate issue. This specific issue shouldn't involve changing `createRule` itself.

For specifically `consistent-type-assertions`, I think this should be doable by removing the schema requirement for `assertionStyle` to be defined:

https://github.com/typescript-eslint/typescript-eslint/blob/3f02687d964c1e7888e5db62256d56cd72c4f281/packages/eslint-plugin/src/rules/consistent-type-assertions.ts#L85

**Comment by scottohara** [comment]
Thanks @JoshuaKGoldberg , to be honest I was starting to get a bit stuck on how to solve this holistically, given Brad's comments in #10339 about not wanting to mess with `context.options` at all.

I think for now I'll just send a PR for `consistent-type-assertions` that, as you suggest, relaxes the schema requirement. 

Thanks for the steer.

**Comment by JoshuaKGoldberg** [comment]
Awesome! And whew, I love a good deep dive but glad you're unblocked. 😅

**Comment by bradzacher** [comment]
> It sounds like I should look further into what makes rules like prefer-destructuring incompatible with the change

Some rules have... complicated schemas and complicated defaults and so a naive "recursive merge" strategy does not work for them because it easily leads to a config that fails schema validation. `prefer-destructuring` is one such example.

This is why I think we really want to avoid adopting upstream `defaultOptions` internally until our minimum version matches the version it was released with -- because there will be differences between what we built ourselves years ago and what eslint chose to build.
